rule peak_annotation:
    input:
        filtered_peaks=lambda wildcards: f"{config['peak_annotation']['input']['filtered_peaks']}/{wildcards.sample}_filtered_peaks.bed"
   
    output:
        annotation=f"{config['peak_annotation']['output']}/{{sample}}_peak_annotation.txt",
        summary=f"{config['peak_annotation']['output']}/{{sample}}_peak_annotation_summary.txt"
        
    params:
        gff=config['peak_annotation']['params']['gff'],
        genome=config['peak_annotation']['params']['genome'],
        feature_types=config['peak_annotation']['params'].get('feature_types', "gene,exon,CDS")
       
    resources:
        mem_mb=config['peak_annotation']['resources']['mem_mb'],
        time=config['peak_annotation']['resources']['time']
                 
    threads:
        config['peak_annotation']['threads']
        
    benchmark:
        "benchmarks/peak_annotation/{sample}.txt"
        
    log:
        "logs/peak_annotation/{sample}.err"
        
    conda:
        "envs/05_analysis/chipseeker.yaml"
        
    message:
        "[Peak annotation] Sample: {wildcards.sample} | Peaks: {input.filtered_peaks} | Output: {output.annotation}"
        
    shell:
        """
        Rscript -e ' \
        library(ChIPseeker); \
        library(GenomicFeatures);\
        
        peakfile <- "{input.filtered_peaks}"; \
        
        txdb <- makeTxDbFromGFF("{params.gff}", format="gtf"); \
        peakAnno <- annotatePeak(peakfile, TxDb=txdb, tssRegion=c(-3000, 3000), verbose=FALSE); \
        
        #Save detailed annotation
        write.table(as.data.frame(peakAnno), "{output.annotation}", sep="\t", row.names=FALSE, quote=FALSE); \
        
        #Save summary counts per feature
        feature_summary <- as.data.frame(table(peakAnno@anno$annotation)); \
        write.table(feature_summary, "{output.summary}", sep="\t", row.names=FALSE, quote=FALSE)' \
        2> {log}
        """
        
        

